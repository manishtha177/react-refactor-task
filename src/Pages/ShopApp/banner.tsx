import React from "react";
import img1 from "../../images/img1.png";
import img2 from "../../images/img2.png";
import styles from "./index.module.css";

const Banner: React.FC = (): JSX.Element => {
  return (
    <>
      <span className={`container ${styles.main} ${styles.images}`}>
        <img data-testid={"img-1"} src={img1} alt="doctor" />
        <img data-testid={"img-2"} src={img2} alt="engineer" />
      </span>
    </>
  );
};

export default Banner;
