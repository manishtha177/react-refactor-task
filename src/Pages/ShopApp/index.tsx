import React, { useState, useEffect } from "react";
import AddProductModal from "../../components/AddProductModal";
import useFetch from "../../hooks/useFetch";
import { IShopProps, IShopState } from "../../interfaces/shop";
import Header from "../../components/Header";
import Banner from "./banner";
import ProductContainer from "../../components/ProductContainer";

const ShopApp: React.FC<IShopProps> = (): JSX.Element => {
  const [shopData, setShopData] = useState<IShopState>({
    products: [],
    isOpen: false,
    isShowingMessage: false,
    message: "",
  });
  const [data] = useFetch("products");
  const setData = (data: IShopProps) => {
    setShopData((prevData) => ({
      ...prevData,
      ...data,
    }));
  };

  const onFavClick = (id: number): void => {
    const prods = shopData?.products;
    const idx = prods?.findIndex((val) => val?.id === id);
    prods[idx].isFavorite = !prods?.[idx]?.isFavorite;
    setData({ products: prods });
  };

  const toggleAddProductModal = (): void => {
    setData({ isOpen: !shopData?.isOpen });
  };

  useEffect(() => {
    setData({ products: data?.reverse() });
  }, [data]);

  return (
    <>
      <Header />
      <Banner />
      <ProductContainer
        shopData={shopData}
        toggleAddProductModal={toggleAddProductModal}
        onFavClick={onFavClick}
      />
      {shopData?.isOpen && (
        <AddProductModal
          isOpen={shopData?.isOpen}
          toggleAddProductModal={toggleAddProductModal}
          setData={setData}
          shopData={shopData}
        />
      )}
    </>
  );
};

export default ShopApp;
