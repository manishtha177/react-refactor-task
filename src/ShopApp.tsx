import React, { useState, useEffect } from "react";
import lodash from "lodash";
import { Button } from "./components/Button";
import ProductList from "./components/ProductList";
import logo from "./images/droppe-logo.png";
import img1 from "./images/img1.png";
import img2 from "./images/img2.png";
import styles from "./shopApp.module.css";
import AddProduct from "./components/AddProduct";
import useFetch from "./hooks/useFetch";
import { IShopProps, IShopState } from "./interfaces/shop";

const ShopApp: React.FC<IShopProps> = () => {
  const [shopData, setShopData] = useState<IShopState>({
    products: [],
    isOpen: false,
    isShowingMessage: false,
    message: "",
  });

  const { fetchProducts, addProduct } = useFetch();

  useEffect(() => {
    getProducts();
    // eslint-disable-next-line
  }, []);

  const setData = (data: IShopProps) => {
    setShopData((prevData) => ({
      ...prevData,
      ...data,
    }));
  };

  const getProducts = async () => {
    const response = await fetchProducts();
    setData({ products: response.reverse() });
  };

  const onFavClick = (id: number) => {
    const prods = shopData.products;
    const idx = lodash.findIndex(prods, { id: id });
    prods[idx].isFavorite = !prods[idx].isFavorite;
    setData({ products: prods });
  };

  const onSubmit = async (payload: {
    title: string;
    description: string;
    price: number;
  }) => {
    setData({
      isOpen: false,
      isShowingMessage: true,
      message: "Adding Product...",
    });

    try {
      const response = await addProduct(payload);
      if (response?.id) {
        const tempProducts = shopData.products;
        tempProducts.unshift({
          id: response?.id,
          title: payload.title,
          description: payload.description,
          price: payload.price,
          isFavorite: false,
          rating: { rate: 0, count: 0 },
        });
        setData({
          products: tempProducts,
          isShowingMessage: true,
          message: "Product added successfully...",
        });
      } else {
        setData({
          isOpen: false,
          isShowingMessage: true,
          message: "Failed to add product...",
        });
      }
    } catch (error) {
      console.log("error : ", error);
      setData({
        isOpen: false,
        isShowingMessage: true,
        message: "Something went wrong",
      });
    } finally {
      setTimeout(() => {
        setData({ isShowingMessage: false, message: "" });
      }, 2000);
    }
  };

  const toggleAddProductModal = () => {
    setData({ isOpen: !shopData.isOpen });
  };

  return (
    <React.Fragment>
      <div data-testid={"header"} className={styles.header}>
        <div className={`container ${styles.headerImageWrapper}`}>
          <img alt="logo" src={logo} className={styles.headerImage} />
        </div>
      </div>

      <>
        <span className={`container ${styles.main} ${styles.images}`}>
          <img data-testid={"img-1"} src={img1} alt="docter" />
          <img data-testid={"img-2"} src={img2} alt="engineer" />
        </span>
      </>

      <div className={`container ${styles.main}`} style={{ paddingTop: 0 }}>
        <div className={styles.buttonWrapper}>
          <span role="button">
            <Button onClick={toggleAddProductModal}>
              Send product proposal
            </Button>
          </span>
          {shopData.isShowingMessage && (
            <div className={styles.messageContainer}>
              <i>{shopData.message}</i>
            </div>
          )}
        </div>

        <div className={styles.statsContainer}>
          <span>Total products: {shopData.products.length}</span>
          {" - "}
          <span>
            Number of favorites:{" "}
            {shopData.products.filter((product) => product.isFavorite).length}
          </span>
        </div>

        {shopData.products && shopData.products.length ? (
          <ProductList products={shopData.products} onFav={onFavClick} />
        ) : (
          <div className={styles.notFound}>No Data Found</div>
        )}
      </div>

      {shopData.isOpen && (
        <AddProduct
          isOpen={shopData.isOpen}
          toggleAddProductModal={toggleAddProductModal}
          onSubmit={onSubmit}
        />
      )}
    </React.Fragment>
  );
};

export default ShopApp;
