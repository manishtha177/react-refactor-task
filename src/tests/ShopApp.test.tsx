import * as React from "react";
import { render, screen } from "@testing-library/react";
import ShopApp from "../Pages/ShopApp";

test("render header & take snapshot for correct render", () => {
  render(<ShopApp />);
  expect(screen.getByTestId("header")).toMatchSnapshot();
  expect(screen.getByTestId("header")).toBeTruthy();
});

test("render image 1", () => {
  render(<ShopApp />);
  expect(screen.getByTestId("img-1")).toBeTruthy();
});

test("render image 2", () => {
  render(<ShopApp />);
  expect(screen.getByTestId("img-2")).toBeTruthy();
});

test("render Send product proposal button", () => {
  render(<ShopApp />);
  expect(screen.getByTestId("send-product-button")).toBeTruthy();
});
