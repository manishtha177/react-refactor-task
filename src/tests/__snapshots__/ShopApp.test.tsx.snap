// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render header & take snapshot for correct render 1`] = `
<div
  class="header"
  data-testid="header"
>
  <div
    class="container headerImageWrapper"
  >
    <img
      alt="logo"
      class="headerImage"
      src="droppe-logo.png"
    />
  </div>
</div>
`;
